python/pyspark/sql/observation.py

```diff
@@ -0,0 +1,147 @@
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from uuid import uuid4
+
+from pyspark.sql import column, Column, DataFrame, Row
+
+basestring = unicode = str
+
+__all__ = ["Observation"]
+
+
+class Observation:
+    """Class to observe (named) metrics on a :class:`DataFrame`.
+
+    Metrics are aggregation expressions, which are applied to the DataFrame while is is being
+    processed by an action.
+
+    .. versionadded:: 3.3.0
+
+    The metrics have the following guarantees:
+
+    - It will compute the defined aggregates (metrics) on all the data that is flowing through
+      the Dataset during the action.
+    - It will report the value of the defined aggregate columns as soon as we reach the end of
+      the action.
+
+    The metrics columns must either contain a literal (e.g. lit(42)), or should contain one or
+    more aggregate functions (e.g. sum(a) or sum(a + b) + avg(c) - lit(1)). Expressions that
+    contain references to the input Dataset's columns must always be wrapped in an aggregate
+    function.
+
+    An Observation instance collects the metrics while the first action is executed. Subsequent
+    actions do not modify the metrics returned by `Observation.get`. Retrieval of the metric via
+    `Observation.get` blocks until the first action has finished and metrics become available.
+
+    This class does not support streaming datasets.
+
+    Examples
+    --------
+    >>> from pyspark.sql.functions import col, count, lit, max
+    >>> from pyspark.sql.observation import Observation
+    >>> observation = Observation("my_metrics")
+    >>> observed_df = df.observe(observation, count(lit(1)).alias("count"), max(col("age")))
+    >>> observed_df.count()
+    2
+    >>> observation.get
+    Row(count=2, max(age)=5)
+    """
+    def __init__(self, name=None):
+        """Constructs a named or unnamed Observation instance.
+
+        Parameters
+        ----------
+        name : str, optional
+            default is a random UUID string. This is the name of the Observation and the metric.
+        """
+        assert isinstance(name, basestring), "name should be a string"
+        self._name = name or str(uuid4())
+        self._jvm = None
+        self._jo = None
+
+    def _on(self, df, *exprs):
```

Comment:
can we keep the same signature as Scala side? e.g.) `_on(self, df, expr, *exprs)` to disallow empty expr

Reply:
The method tests for empty exprs, but the Scala signature checks this on compile time, which is better.

I have aligned that with existing `agg` method, which has a single `exprs` and checks for non-emptiness.

Do you still say to mimic the Scala signature? Would you even say DataFrame.agg should do the same? I could draft another PR to fix that for agg if you let me. That shouldn't break python user code, right?

Reply:
Hm, okay. let's leave it as is for now.